File: Práctica 1/Java/Coleccion.java
//23900715N HERNANDEZ GOMEZ, JESSICA
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.*;

public class Coleccion {
	private ArrayList<ArrayList<Character>> mapa;
	private ArrayList<Localidad> localidades;

	public Coleccion() {
		mapa = new ArrayList<ArrayList<Character>>();
		localidades = new ArrayList<Localidad>();
	}

	public void lectura(String f) {
		FileReader fichero = null;
		BufferedReader lectura = null;
		String linea;
		String[] elementos;
		String separador = "[ ]+";
		Localidad l;
		InfoTur it;

		int museo = -1;
		int monumento = -1;
		int hotel = -1;
		int restaurante = -1;
		boolean ae = false;
		String top = "";

		try {
			fichero = new FileReader(f);
			lectura = new BufferedReader(fichero);
			linea = lectura.readLine();

			int x = 0;
			while (linea != null) {
				if (linea.equals("<LOCALIDAD>") == false) {
					// LECTURA MAPA
					//System.out.println("Empiezo a leer el mapa");
					mapa.add(new ArrayList<Character>());
					
					for (int y = 0; y < linea.length(); y++) {
						mapa.get(x).add(linea.charAt(y));
					}
					
					linea = lectura.readLine();
					x++;
					//System.out.println("He terminado de leer el mapa");
				} else {
					// LECTURA LOCALIDADES
					//System.out.println("Empiezo a leer localidades");
					linea = lectura.readLine(); // Nombre localidad
					l = new Localidad(linea); // Creo la localidad
					linea = lectura.readLine();
					elementos = linea.split(separador);
					//System.out.println("Elemento 1: " + Integer.parseInt(elementos[0]) + "Elemento 2: " + Integer.parseInt(elementos[1]));
					int esta=l.setCoor(Integer.parseInt(elementos[0]), Integer.parseInt(elementos[1]), mapa) ;//!= -1){ // Agrego coordenadas
						// INFORMACION
						//System.out.println("Empiezo con la info de las localidades");
						linea = lectura.readLine(); // <INFO>
						linea = lectura.readLine();
						while (linea!=null && linea.equals("<LOCALIDAD>") == false) {
							elementos = linea.split(separador);
							if (elementos[0].equals("museo")) {
								museo = Integer.parseInt(elementos[1]);
							}
							if (elementos[0].equals("monumento")) {
								monumento = Integer.parseInt(elementos[1]);
							}
							if (elementos[0].equals("hotel")) {
								hotel = Integer.parseInt(elementos[1]);
							}
							if (elementos[0].equals("restaurante")) {
								restaurante = Integer.parseInt(elementos[1]);
							}
							if (elementos[0].equals("aeropuerto")) {
								ae = true;
							}
							if (elementos[0].contains("*")) {
								top = linea;
							}
							linea = lectura.readLine();
						}
	
						it = new InfoTur(museo, monumento, hotel, restaurante, ae);
						it.setTop(top);						
						museo = -1;
						monumento = -1;
						hotel = -1;
						restaurante = -1;
						ae = false;
						top = "";
						l.setInfo(it);
						if(esta>-1)
						localidades.add(l);
						
					}
				}
			
		} catch (IOException e) {
			System.err.println("Error con el fichero");
			System.exit(0);
		}

		try {
			if (fichero != null) {
				fichero.close();
			}
			if (lectura != null) {
				lectura.close();
			}
		} catch (IOException ex) {
			System.out.println(ex);
		}
	}

	public ArrayList<ArrayList<Character>> getMapa() {
		return mapa;
	}

	public ArrayList<Localidad> getLocalidades() {
		return localidades;
	}

	public char getCoorMapa(Coordenadas c) {
		char devolver = 'X';
		
		if(c.getFila() >= 0 && c.getColumna() >= 0){
			if(c.getFila() < mapa.size() && c.getColumna() < mapa.get(0).size()){
				devolver = mapa.get(c.getFila()).get(c.getColumna());
			}
		}
		return devolver;
	}

	public String toString() {
		String devolver = "";
		// MAPA
		for (int x = 0; x < mapa.size(); x++) {
			for (int y = 0; y < mapa.get(x).size(); y++) {
				devolver = devolver + mapa.get(x).get(y);
			}
			devolver = devolver + "\n";
		}
		// LOCALIDADES
		for (int z = 0; z < localidades.size(); z++) {
			devolver = devolver + localidades.get(z).toString();
		}

		return devolver;
	}
}

File: Práctica 1/C++/Coleccion.cc
//DNI 23900715N HERNANDEZ GOMEZ, JESSICA
#include "Coleccion.h"
#include <sstream>

ostream & operator<< (ostream &os, const Coleccion &c) { //Funcion amiga de sobrecarga del operador salida
    //MAPA
    for(unsigned x = 0; x<c.mapa.size(); x++){
        for(unsigned y = 0; y<c.mapa[0].size(); y++){
            os << c.mapa[x][y];
        }
        os << endl;
    }
    //LOCALIDADES
    for(unsigned int z = 0; z<c.localidades.size(); z++){
        os << c.localidades[z];
    }
    return os;
}

Coleccion::Coleccion(){

}

Coleccion::Coleccion(const Coleccion &c){
    mapa = c.mapa;
    localidades = c.localidades;
}

Coleccion::~Coleccion(){
    mapa.clear();
    localidades.clear();
}

Coleccion &Coleccion::operator=(const Coleccion &c){
    mapa = c.mapa;
    localidades = c.localidades;
    return *this;
}

void Coleccion::lectura(const string f){
    fstream fichero;
    fichero.open(f.c_str());
    string linea;
    int esta = -1;
    vector<int> coords;
    bool espacio = false;
    int coordX = -1;
    int coordY = -1;
    string scoordX = "";
    string scoordY = "";
    string info = "";
    int museo = -1;
    int monumento = -1;
    int hotel = -1;
    int restaurante = -1;
    bool ae = false;
    string top = "";
    vector<char> v;
    
    if(fichero.is_open()){
        while(getline(fichero, linea)&&linea!="<LOCALIDAD>"){
            //LECTURA MAPA
            for (unsigned y = 0; y<linea.length(); y++){
                v.push_back(linea[y]);
            }
            mapa.push_back(v);
            v.clear();
        }
        while(getline(fichero, linea)){
            //LECTURA LOCALIDADES
            Localidad l(linea); //Creo la localidad
            getline(fichero, linea);
            //Separo las coordenadas
            for (unsigned x = 0; x<linea.length(); x++){
                if(linea[x] == ' '){
                    espacio = true;
                }
                if(!espacio){
                    scoordX = scoordX + linea[x];
                }else if(espacio){
                    scoordY = scoordY + linea[x];
                }
            }
            espacio = false;

            istringstream(scoordX)>>coordX;
            istringstream(scoordY)>>coordY;
            esta = l.setCoor(coordX, coordY, getMapa());//Agrego coordenadas
            scoordX.clear();
            scoordY.clear();
            //INFORMACION
            getline(fichero, linea); //<INFO>
            getline(fichero, linea);
            while(linea != "<LOCALIDAD>" && !fichero.eof()){
                if(linea[0] == 'm'){//Caso Museo y Monumento
                    if(linea[1] == 'u'){ //Museo
                        stringstream ss(linea);
                        ss>>info;
                        ss>>museo;
                    }else{
                        stringstream ss(linea); // Monumento
                        ss>>info;
                        ss>>monumento;
                    }
                }
                if(linea[0] == 'h'){
                    stringstream ss(linea);
                    ss>>info;
                    ss>>hotel;
                }
                if(linea[0] == 'r'){ 
                    stringstream ss(linea);
                    ss>>info;
                    ss>>restaurante;
                }
                if(linea[0] == 'a'){
                    ae = true;
                }
                if(linea[0] == '*'){
                    top = linea;
                }
                getline(fichero, linea);
            }
            InfoTur it = InfoTur(museo, monumento, hotel, restaurante, ae);
            it.setTop(top);
            museo = -1;
            monumento = -1;
            hotel = -1;
            restaurante = -1;
            ae = false;
            top.clear();

            l.setInfo(it);
            if(esta > -1){
                localidades.push_back(l);
            }
        }
        fichero.close();
    }
}

vector<vector<char> > & Coleccion::getMapa(){
    return mapa;
}

vector<Localidad> &Coleccion::getLocalidades(){
    return localidades;
}

char Coleccion::getCoorMapa(Coordenadas c){
    char devolver = 'X';
    int f = c.getFila();
    int coor = c.getColumna();

    for(unsigned int x = 0; x<mapa.size(); x++){
        for(unsigned int y = 0; y<mapa[x].size(); y++){
            if(f == (int)x && coor == (int)y){
                devolver = mapa[x][y];
            }
        }
    }
    return devolver;
}